[WIP] Add documentation on build systems

adamjstewart opened the pull request:
Adds documentation for the following build systems:

- [x] `MakefilePackage`
- [ ] `AutotoolsPackage`
- [ ] `CMakePackage` - waiting on #4988
- [x] `QMakePackage`
- [ ] `PythonPackage`
- [ ] `RPackage`
- [ ] `PerlPackage`
- [ ] `RubyPackage` - waiting on #3127
- [ ] `SConsPackage`
- [ ] `WafPackage`
- [ ] `IntelPackage` - waiting on #4300
- [ ] Custom build systems

We never really got around to documenting how to use each build system when we introduced separate base classes. This PR is an attempt to solve this. It is also an attempt to pass on knowledge of each build system to new package developers who may not be familiar with SCons or QMake or Waf. If there's anything you think I'm missing, let me know!

alalazo commented:
:100: I had a first look at I think this is great! Let me know when this is ready to be reviewed.

adamjstewart commented:
Unfortunately, grad school has kept me much too busy to work on this. I would like to see this get merged before SC17 though. If anyone has the time to contribute to this PR, feel free to add information on the build systems I didn't get to. The structure of what I had intended should be pretty obvious from the other build systems.

adamjstewart commented:
@davydden Now that #6746 has been merged, would you mind adding a page on `OctavePackage` to this PR while it's still fresh in memory?

davydden commented:
@adamjstewart 

> Now that #6746 has been merged, would you mind adding a page on OctavePackage to this PR while it's still fresh in memory?

sure, see https://github.com/adamjstewart/spack/pull/4

adamjstewart commented:
@hartzell If you get a chance, would you mind adding a page for `AspellDictPackage`? I think I'll try to finish this PR up before the semester starts.

adamjstewart commented on lib/spack/docs/build_systems/rpackage.rst:
```diff
@@ -0,0 +1,338 @@
+distributed via CRAN. Because they're distributed via CRAN, they can also be
+updated between R version releases.
+
+Spack explicitly adds the ``--without-recommended-packages`` flag to prevent
```
This is not actually true at the moment, the PR that introduced this was abandoned due to merge conflicts. We really need to make this happen if we want reliable R activation support.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+``--without-baz``. Since these flags are so common, Spack provides a
+couple of helper functions to make your life easier.
+
+TODO: document ``with_or_without`` and ``enable_or_disable``.
```
I must admit, I don't actually know the full extent of how these helper functions work. @alalazo would you be willing to write this section of the docs?

adamjstewart replied:
> If these helper functions are considered best practice, please use them in your example above.

I was hoping that the example above would be a perfect segue into the "Helper functions" section. Like, "in the previous example, we used this long syntax. It's pretty common, so here's a shorter way of doing the same thing.

I agree that too much magic is a bad thing, but it does currently exist in Spack. I would err on the side of documenting it unless there is a consensus that this is too much magic. In this case, we should remove it instead of keeping it around.

citibeth replied:
The helper functions are not a deep idea, I don't think they need their own section.

adamjstewart commented:
Alright, I believe this PR is officially ready for review! I think I added everything I know, but I know many of you have much more experience with Autotools/CMake than I do, so please correct me if I made any mistakes or neglected to mention any important details.

We don't have docs for `RubyPackage` yet, as it hasn't been merged yet. Similarly, `IntelPackage` is currently in a state of flux, so I'm not too worried about adding docs until after #7469 is merged. I'm not particularly worried about `AspellDictPackage`, as it isn't really a build system. Honestly, I don't think it belongs in the build systems category, we ought to move it elsewhere someday.

@svenevs you might be interested in this PR if you have the time to review. I know you're just as opinionated as I am when it comes to the _right_ way of doing things 😄

adamjstewart commented:
P.S. If you find anything in the Packaging Guide that feels build system-specific, let me know and I'll move it to these docs. I'm adding over 3,000 lines of documentation, so I want to keep it succinct as possible. If we duplicate things in the Packaging Guide too, no one will ever want to read it 😛

davydden commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+       env['BLASLIB'] = spec['blas'].libs.ld_flags
+
+
+``cbench`` is a good example of a simple package that does this, while
```
Can u please add url links in case users would like to have a look?

adamjstewart replied:
I was hoping that people would just use `spack edit`...

citibeth replied:
All we know that people will have when they read this is a web browser.  They might not yet have Spack set up.  Or they might, but they're not logged onto that machine.  Or they work in an environment where `spack edit` doesn't work for them.   Not everyone is going to use `spack edit`, and the docs should not be biased toward trying to force them to do so.

citibeth commented on lib/spack/docs/build_systems/cmakepackage.rst:
```diff
@@ -0,0 +1,258 @@
+
+   variant('build_type', default='DebugRelease',
+           description='The build type to build',
+           values=('Debug', 'Release', 'DebugRelease'))
```
Should dealii's `DebugRelease` be changed to `RelWithDebInfo`?  Why is `dealii` nonstandard anyway?  Can we remove this?

adamjstewart replied:
Unfortunately, `dealii` isn't the only one:
```console
$ grep 'variant(.*build_type' */package.py
benchmark/package.py:    variant('build_type', default='RelWithDebInfo',
cbtf-argonavis/package.py:    variant('build_type', default='None', values=('None'),
cbtf-krell/package.py:    variant('build_type', default='None', values=('None'),
cbtf-lanl/package.py:    variant('build_type', default='None', values=('None'),
cbtf/package.py:    variant('build_type', default='None', values=('None'),
cgal/package.py:    variant('build_type', default='Release',
dealii/package.py:    variant('build_type', default='DebugRelease',
eccodes/package.py:    variant('build_type', default='RelWithDebInfo',
eigen/package.py:    variant('build_type', default='RelWithDebInfo',
elemental/package.py:    variant('build_type', default='Release',
fenics/package.py:    variant('build_type', default='RelWithDebInfo',
grib-api/package.py:    variant('build_type', default='RelWithDebInfo',
gromacs/package.py:    variant('build_type', default='RelWithDebInfo',
hydrogen/package.py:    variant('build_type', default='Release',
jsoncpp/package.py:    variant('build_type', default='RelWithDebInfo',
krims/package.py:    variant("build_type", default="DebugRelease",
lazyten/package.py:    variant("build_type", default="DebugRelease",
lbann/package.py:    variant('build_type', default='Release',
libemos/package.py:    variant('build_type', default='RelWithDebInfo',
llvm-openmp-ompt/package.py:    variant('build_type', default='Release',
llvm/package.py:    variant('build_type', default='Release',
magics/package.py:    variant('build_type', default='RelWithDebInfo',
mbedtls/package.py:    variant('build_type', default='Release',
metis/package.py:    variant('build_type', default='Release',
opencoarrays/package.py:    variant('build_type', default='RelWithDebInfo',
openspeedshop/package.py:    variant('build_type', default='None', values=('None'),
psi4/package.py:    variant('build_type', default='Release',
relion/package.py:    variant('build_type', default='RelWithDebInfo',
rr/package.py:    variant('build_type', default='Release',
symengine/package.py:    variant('build_type', default='Release',
tethex/package.py:    variant('build_type', default='Release',
typhonio/package.py:    variant('build_type', default='Release', description='The build type to build',
vc/package.py:    variant('build_type', default='RelWithDebInfo',
```
Maybe @davydden can comment on this.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+MakefilePackage
+---------------
+
+The most primitive build system a package can use is a plain Makefile.
```
Not true.  The most primitive is a script you have to edit.  Believe me, it still exists...

adamjstewart replied:
Fair enough, although I'm not sure if that's considered a "build system" or not.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+
+``Makefile`` is the most common name.
+
+Some Makefiles also *include* other configuration files. Check for an
```
Included files typically end in `.mk`.

adamjstewart replied:
Not sure if I've actually seen a `.mk` file before. I see `.inc` a lot though.

citibeth replied:
I've seen `.mk`

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+Make has multiple types of assignment operators. Some Makefiles
+use ``=`` to assign variables. The only way to override these
+variables is to edit the Makefile or override them on the command-line.
+However, Makefiles that use ``?=`` for assignment honor environment
```
The sentence above is irrelevant to "Environment Variables" and should be moved to a different section.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+.. code-block:: python
+
+   def edit(self, spec, prefix):
+       env['PREFIX'] = prefix
```
Where does the `env` variable come from?  Why is this different from `setup_environment()`?

adamjstewart replied:
> Where does the `env` variable come from?

https://github.com/spack/spack/blob/develop/lib/spack/spack/build_environment.py#L364

> Why is this different from `setup_environment()`?

The only real difference I've seen is that `setup_environment` is run before `edit`, and before we switch to the build directory. So things that need to know the current build directory will fail unless you set them in `edit`.

citibeth replied:
This information needs to be included in the docs.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+
+       return [
+           'CC=cc',
+           'BLASLIB={0}'.format(spec['blas'].libs.ld_flags),
```
I recommend using "%s"%xxx instead of "{0}".format() because (a) it fits better on flake8 lines, and (b) '%s' will always be supported in Python, it's not going away.

adamjstewart replied:
`%s` has many known problems. For instance:

- It only supports ints, strings, and doubles. All other types (including the `Prefix` type) must be converted to a string first.
- `%` is a binary operator, and can only take 2 arguments.

This is why [PEP 3101](https://www.python.org/dev/peps/pep-3101) was introduced. In fact, if you read the [string formatting docs](https://docs.python.org/3.6/library/string.html), `%s` isn't even mentioned. There are currently no plans to remove it, but `.format()` is far more powerful, and supports nesting of replacement variables among other things. Also see [PEP 498](https://www.python.org/dev/peps/pep-0498/) for the future of string formatting in Python.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+
+.. code-block:: python
+
+   build_targets = ['CC=cc']
```
This doesn't make sense to me.  Why does `build_targets` set command-line variables?  What `make` command line is ultimately generated by the way this is set?

adamjstewart replied:
`build_targets` is simply added to the `make()` call. I know these aren't really targets, but people have used them for this purpose in other packages.

citibeth replied:
Ahh, that makes more sense.  It is totally an abuse of the feature, so that needs to be explained.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+       makefile = FileFilter('Makefile')
+
+       makefile.filter('CC = gcc',  'CC = cc')
+       makefile.filter('CXX = g++', 'CC = c++')
```
It seems inefficient to process makefile once per filter.  Is there a way that multiple filters could be applied in one pass?

Also.. the variable here should probably be called `Makefile` because that's what the file is called.

adamjstewart replied:
`filter` only supports a single search and replace argument at the moment.

For the variable name, see [PEP 8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names).

citibeth replied:
The variable is technically a constant because it's set once and never changed.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+
+
+``hpl`` is a good example of a package that uses a list to store
+configuration variables.
```
Should we avoid the use of Python's OrderedDict?  What is Spack's policy on this?

adamjstewart replied:
`OrderedDict` is [new in Python 2.7](https://docs.python.org/2/library/collections.html). Spack has its own `OrderedDict` backport, but I think that's overkill for this job.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+
+  This includes variables such as ``CC``, ``CXX``, ``F77``, ``F90``,
+  and ``FC``, as well as variables related to MPI compiler wrappers,
+  like ``MPICC`` and friends.
```
Instead of "and friends", just list them out.  I know it's tedious, but it makes the docs more useful.

citibeth commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+  ``CXXFLAGS``, ``F77FLAGS``, ``F90FLAGS``, ``FCFLAGS``, and ``CPPFLAGS``.
+  These variables are often hard-coded to contain flags specific to a
+  certain compiler. If these flags don't work for every compiler,
+  you may want to consider filtering them.
```
What do you mean by "filter"?  Edit?

adamjstewart replied:
Replace them using the `filter` command described above

citibeth commented on lib/spack/docs/build_systems/perlpackage.rst:
```diff
@@ -0,0 +1,209 @@
+Obviously, these commands can only be run if you have root privileges.
+Furthermore, ``cpanm`` is not capable of installing non-Perl dependencies.
+If you need to install to your home directory or need to install a module
+with non-Perl dependencies, Spack is a better option.
```
Mention that you can install Perl with Spack, and then use `cpanm` from that Perl to install packages from the internet, no root required.  This violates a core assumption of Spack (two installed packages with the same spec are the same); but it's still a useful thing to do in some cases.

Also, you can invent new Spack "versions" of Perl to use for installing all your stuff, while leaving a "standard" Perl for use by Spack packages.

The same advice goes for Python, Nodejs, R, etc.

scheibelp replied:
That sounds useful to mention but not essential for this PR. Views may allow a user to symlink a package into a new prefix, then install addons without modifying the Spack installation prefix, (e.g. this is possible for python with virtualenv once https://github.com/spack/spack/pull/7152 is merged).

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+* bdist_rpm
+* bdist_wininst
+* upload
+* check
```
No way... why so many?

adamjstewart replied:
See the next sentence.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+
+As you can imagine, it is impossible for Spack to guess what URL to
+use to download version 1.12.0 given this URL. There is a solution,
+however. PyPI offers a pre-production interface for downloading
```
What does "pre-production" mean herre?

adamjstewart replied:
When I wrote this, `pypi.io` was a "pre-production" warehouse, meaning that it was in development and not yet stable. Since then, it looks like `pypi.io` may be official now. It no longer has its own website.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+PyPI vs. GitHub
+"""""""""""""""
+
+Many packages are hosted on PyPI, but are developed on GitHub. The tarball
```
GitHub (or elsewhere).

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+Build system dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There are a few dependencies common to the PythonPackage build system.
```
Use `PythonPackage`

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+
+When you install a package with ``pip``, it reads the ``setup.py``
+file in order to determine the dependencies of the package.
+If the dependencies are not yet installed, ``pip`` downloads them
```
I don't understand the purpose of this section.  Are you telling the reader to NOT use PIP  to install within Spack?  Or... how to avoid auto-downloads for Python packages that auto-download with pip?

adamjstewart replied:
`python setup.py` will auto-download these dependencies if they aren't already installed.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There are many packages that make use of Python, but packages that depend
+on Python are not necessarily PythonPackages.
```
...why?  Because they don't use `setup.py`!  I've made some upstream software that is mostly C++ and builds with CMake --- even the Python portions.
In this case, you need to say `extends('python')`.  Note to avoid `extends('python@3:')`, instead do 
```
extends('python')
depends_on('python@3:')
```

adamjstewart replied:
See ... this section.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+library that will be imported in other Python scripts? Or is it a
+command-line tool, or C/C++/Fortran program with optional Python
+modules? The former should be prepended with ``py-``, while the
+latter should not.
```
Please address the issue.... if the upstream package is called `PyXyz`, should you call your package `py-xyz` or `py-pyxyz`?

adamjstewart replied:
It should be called `py-pyxyz`. For example:
```console
$ spack list py-py
==> 67 packages.
py-py          py-pybind11         py-pydispatcher  py-pygobject   py-pypar        py-pyside          py-pytest-runner       py-pyutilib
py-py2bit      py-pybtex           py-pydot         py-pygpu       py-pyparsing    py-pysocks         py-pytest-xdist        py-pyvcf
py-py2cairo    py-pybtex-docutils  py-pyelftools    py-pygtk       py-pypeflow     py-pysqlite        py-python-daemon       py-pywavelets
py-py2neo      py-pychecker        py-pyfaidx       py-pylint      py-pyprof2html  py-pytables        py-python-engineio     py-pyyaml
py-py4j        py-pycodestyle      py-pyfasta       py-pymatgen    py-pyqt         py-pytest          py-python-gitlab
py-pyani       py-pycparser        py-pyfftw        py-pyminifier  py-pyrad        py-pytest-cov      py-python-levenshtein
py-pyasn1      py-pycrypto         py-pyflakes      py-pymol       py-pysam        py-pytest-flake8   py-python-socketio
py-pybedtools  py-pycurl           py-pygdbmi       py-pympler     py-pyscaf       py-pytest-httpbin  py-pythonqwt
py-pybigwig    py-pydatalog        py-pygments      py-pynn        py-pyserial     py-pytest-mock     py-pytz
```

citibeth replied:
> spack create handles this anyway, so it shouldn't really be a problem.

Not everybody uses `spack create`.

adamjstewart replied:
No need to manually change it, you can run `spack create --template python --force url` to change it.

P.S. If you ever think it isn't correctly identifying a package build system let me know and I can take a look.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+the ability to _activate_ the package. Activation involves symlinking
+everything in the installation prefix of the package to the installation
+prefix of Python. This allows the user to import a Python module without
+having to add that module to ``PYTHONPATH``.
```
Point out that this is not necessary.  You can also just load the module.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+PyPI has hundreds of thousands of packages that are not yet in Spack,
+and ``pip`` may be a perfectly valid alternative to using Spack. The
+main advantage of Spack over ``pip`` is its ability to compile
+non-Python dependencies. It can also build cythonized versions of a
```
...and Spack's multi-versioning / concretization features.  But it's a bit overkill for a pure Python environment because Python is more "flexible" in terms of which versions of A can be used with B.  Spack is a big win when you have significant C/Fortran and Python stuff.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+package or link to an optimized BLAS/LAPACK library like MKL,
+resulting in calculations that run orders of magnitude faster. If you
+only need Python tools like flake8 and sphinx, don't waste your time
+on Spack. But if you need packages with non-Python dependencies like
```
Not true about wasting time.  I'm about to add a `spack-hacking` environment to Spack that allows users to get up-and-running quickly with all the `flake8`, `sphinx`, etc. needed to work with Spack.  Not having to install a completely different and separate system is a real advantage.

citibeth commented on lib/spack/docs/build_systems/pythonpackage.rst:
```diff
@@ -0,0 +1,733 @@
+non-Python dependencies. Anaconda contains many Python packages that
+are not yet in Spack, and Spack contains many Python packages that are
+not yet in Anaconda. The main advantage of Spack over Anaconda is its
+ability to choose a specific compiler and BLAS/LAPACK or MPI library.
```
And its flexible concretization / multi-versioning; which is much more primitive in Conda.  Conda was at first primarily Python, but Spack is more language-neutral.  The big advantage of Conda is it works on Windows.

adamjstewart replied:
@tgamblin keeps flirting with the idea of Windows support 😆

citibeth commented on lib/spack/docs/build_systems/sconspackage.rst:
```diff
@@ -0,0 +1,296 @@
+
+Unlike Autotools and CMake, SCons is a general-purpose build system
+that does not rely on Makefiles to build software. SCons is written
+in Python, and handles all building and linking itself.
```
Think of Scons as a Python-based "make".  The two should be grouped together.

adamjstewart replied:
`scons` and `python` should be grouped together or `scons` and `make`? I grouped things based on https://en.wikipedia.org/wiki/List_of_build_automation_software

citibeth replied:
I think that Wikipedia list leaves a lot to be desired, at least for our purposes.  I would classify them as follows:

1. Dependency management systems (of which make was the first): These know that you have to complete tasks A and B before you can run task C.  They keep track of which tasks have been completed as they work their way through the set of tasks to complete.  scons, ant and make are all fundamentally the same idea here.  The difference is how tasks are packaged and run.  In make, tasks are bash commands.  In ant, tasks are Java classes.  In scons, tasks are Python functions.

The fact that scons is written in Python is immaterial here.  Scons is not a "Python build system" in the way that distutils and setuputils are, meaning they are commonly used to build Python packages.  Scons is really a make replacement, for people who think that Python is a better language for Makefiles than GNU make.  The only time I've seen Scons used is in a C++ shop, to build C++.  It turned out to be really slow.  Eventually, it was ripped out and replaced with Makefiles.  It turns out that the GNU make language is (a) fast, and (b) effective due to its functional language lazy-evaluation nature.  Replacing it with a slow, eager-evaluation language was not a good idea.  In the end, everyone's moving to CMake, which makes the lower-level dependency management system (Make / Ninja / Scons / Ant) a little irrelevant, as long as CMake can generate for it.

2. Script-generation systems: CMake, Autotools.  You have to run a configure procedure BEFORE you can run the depenency DAG from (1) above.

3. Meta-builders: Spack


================
@tgamblin We need to get Spack added to that Wikipedia website under"Meta-build".
https://en.wikipedia.org/wiki/List_of_build_automation_software

citibeth commented on lib/spack/docs/build_systems/wafpackage.rst:
```diff
@@ -0,0 +1,124 @@
+----------
+
+Like SCons, Waf is a general-purpose build system that does not rely
+on Makefiles to build software.
```
What are example packages that use `waf`?

adamjstewart replied:
`py-py2cairo` and `tut` are the only Waf packages currently in Spack.

adamjstewart commented on lib/spack/docs/build_systems.rst:
```diff
@@ -0,0 +1,78 @@
+
+.. code-block:: console
+
+   $ spack edit --build-system autotools
```
Let's not reopen old wounds. This was settled on in #2865. You might not find it useful, but I probably used it hundreds of times while writing these docs. I don't know if I could survive without it.

citibeth replied:
Please just provide appropriate alternatives.  Tell people where the file is located in case they want to edit it themselves, give URLs to GItHub, etc.

adamjstewart commented on lib/spack/docs/build_systems.rst:
```diff
@@ -0,0 +1,78 @@
+   $ grep -l QMakePackage */package.py
+
+
+You can then view these packages with ``spack edit``.
```
```console
$ spack edit foo
```
is fewer characters than
```console
$ less ~/spack/var/spack/repos/builtin/packages/foo/package.py
```
so I went with the former.

citibeth replied:
Again... people need to know where the files really are if they aren't going to see Spack as magic.

@tgamblin I would also LOVE to shorten some of these needlessly long paths in Spack.  I understand the rationale of mirroring a Unix filesystem.  But I don't see how it helps us, or where it is useful.  What I DO see is unnecessarily long paths, along with having to remember which bits of Spack end up in /var vs /opt vs/ share vs /lib.  It all seems so arbitrary and NOT user friendly.

At the very least, I would like symlink shortcuts to commonly used directories.  I asked for this before and the idea was rejected.  But I still want them, and I still add them routinely to my branches.

adamjstewart commented on lib/spack/docs/build_systems/Autoconf-automake-process.svg:
```diff
@@ -0,0 +1,840 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
I didn't write this, I just stole it from Wikipedia. If you want to rewrite it in dot, let me know. Otherwise I'll leave it as is.

citibeth replied:
If you took it from Wikipedia, you need to acknowledge that.

adamjstewart replied:
Done. It looks like the source code to generate the diagram can be found at https://commons.wikimedia.org/w/index.php?curid=15581407 if we want to modify it.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+#. ``install`` - install the package
+
+Most of the time, the ``autoreconf`` phase will do nothing, but if the
+package is missing a ``configure`` script, ``autoreconf`` will generate
```
See the "Build system dependencies" section.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+package is missing a ``configure`` script, ``autoreconf`` will generate
+one for you.
+
+The other phases run:
```
Well, technically, these phases (functions) actually run these commands.

citibeth replied:
It would be helpful if the commands typically run by the phases are labelled by which phase they are run by.  This goes for all the similar segments in the build system docs.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+
+
+Of course, you may need to add a few arguments to the ``./configure``
+line.
```
See the "Finding configure flags" and "Adding flags to configure" sections below.

citibeth replied:
Just saying... I think a brief "why" would be useful here.  Either that, or eliminate "Of course."  Saying "of course" to something that isn't obvious to the reader is probably not a good idea.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+
+The most important file for an Autotools-based package is the ``configure``
+script. This script is automatically generated by Autotools and generates
+the appropriate Makefile when run.
```
This is explained in the next section.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+   steps to install a package. For this reason, some developers will write
+   their own ``configure`` scripts that have nothing to do with Autotools.
+   These packages may not accept the same flags as other Autotools packages,
+   so it is better to create a custom build system. You can tell if a package
```
I'll add a link to the "Custom build system" docs. I'm suggesting they use `Package` and do a one-off.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+``configure.ac`` into ``configure``, while ``automake`` is used to
+convert ``Makefile.am`` into ``Makefile.in``. ``Makefile.in`` is used
+by ``configure`` to generate a platform-dependent ``Makefile`` for you.
+The following diagram provides a high-level overview of the process:
```
See the "External documentation" section below. Let me know if you know of any additional docs you would like me to add.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+   depends_on('autoconf', type='build', when='@develop')
+   depends_on('automake', type='build', when='@develop')
+   depends_on('libtool',  type='build', when='@develop')
+   depends_on('m4',       type='build', when='@develop')
```
Honestly, if we were going to do anything, I would add a subclass of `AutotoolsPackage` called `AutoreconfPackage` (or something like that) that adds these dependencies already. But let's save that for another day. `spack create` already automatically adds these dependencies if it finds `configure.ac` (and friends) but no `configure`.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+Another reason for needing Autotools dependencies is if the developers
+distribute a patch that modifies one of the files used to generate
+``configure`` or ``Makefile.in``. If this is the case, these scripts
+will need to be regenerated. It is preferable to regenerate these
```
Autotools dependencies have been historically buggy on Fedora for some reason, not sure if that has been fixed. Anyway, I definitely prefer it. I've had pretty good luck with running these locally and generating small patches to `configure`. In my mind, it beats installing 4+ additional dependencies any day.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+   force_autoreconf = True
+
+This line tells Spack to wipe away the existing ``configure`` script
+and generate a new one.
```
Idk if 3 deserves its own section, it's only a paragraph at the moment. Let's see if anyone else disagrees on whether we should patch `configure` or `autoreconf`. But I can separate 2 into its own subsection and add an example of how to `force_autoreconf` for a single version only, good point.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+       if '+mpi' in self.spec:
+           args.append('--enable-mpi')
+       else:
+           args.append('--disable-mpi')
```
I actually prefer the former, I find it easier to read. Also, I often run into problems fitting your suggestions under the 80 character limit.

citibeth replied:
Existing and new packages do this in a variety of styles.  The docs should reflect that.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+Note that we are explicitly disabling MPI support if it is not
+requested. This is important, as many Autotools packages will enable
+options by default if the dependencies are found, and disable them
+otherwise. We want Spack installations to be as deterministic as possible.
```
Can you elaborate? I think that's what this paragraph is suggesting already.

citibeth replied:
See "Don't Abuse Automagic Dependencies" https://www.linux.com/news/best-practices-autotools

And here: "Why Automagic Dependencies are Wrong" https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Automagic_dependencies

Maybe we could like to these articles (especially the second), for readers who are interested in more.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+
+.. code-block:: python
+
+   configure_directory = 'src'
```
> show that this is inside a package

Everything in all of these docs is inside a package. I'm not sure if that needs to be said, but if it does, it should probably be somewhere higher up.

> any suggestions on where this goes in order of variants, versions, url, etc?

I like to order things like:

- docstring
- homepage
- url
- list_url
- versions
- variants
- configure_directory
- build_directory

but that's just me. Not sure if this is worth documenting.

citibeth replied:
> Not sure if this is worth documenting.

I've submitted package PR's and had changes requested due to ordering.  If certain orders can be rejected, then acceptable orderings need to be documented.

> Everything in all of these docs is inside a package. 

Adding the context in examples will help newbies.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+   build_directory = 'spack-build'
+
+By default, Spack will build the package in the same directory that
+contains the ``configure`` script
```
We do use it all the time for `CMakePackage`. We don't use it all the time for `AutotoolsPackage` because we were afraid to break things without better testing. If we ever get individual package testing in place, this would be something to consider.

I'm not sure if this would help with #8153. My understanding is that you need access to the source code of `tcl` when building `tk`. I don't think `build_directory` would help here.

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+.. code-block:: python
+
+   build_targets = ['all', 'docs']
+   install_targets = ['install', 'docs']
```
Yes, this would be a perfectly reasonable means of adding `install-headers` and `install-private-headers` to the `make` invocation. You would have to use the:
```python
@property
def install_targets(self):
```
syntax to make it conditional

adamjstewart commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,281 @@
+
+Autotools-based packages typically provide unit testing via the
+``check`` and ``installcheck`` targets. If you build your software
+with ``--test=root``, Spack will check for the presence of a ``check``
```
It means you run `spack install --test=root`. Good point, I'll reword this.

adamjstewart commented on lib/spack/docs/build_systems/cmakepackage.rst:
```diff
@@ -0,0 +1,258 @@
+
+A few more flags are passed to ``cmake`` by default, including flags
+for setting the build type and flags for locating dependencies. Of
+course, you may need to add a few arguments yourself.
```
See the "Finding cmake flags" and "Adding flags to cmake" sections below.

citibeth replied:
Again... don't use "of course" on statements that aren't obvious.  or be prepared to explain why.

adamjstewart commented on lib/spack/docs/build_systems/cmakepackage.rst:
```diff
@@ -0,0 +1,258 @@
+       if '+hdf5' in self.spec:
+           args.append('-DDETECT_HDF5=ON')
+       else:
+           args.append('-DDETECT_HDF5=OFF')
```
See above comment.

adamjstewart commented on lib/spack/docs/build_systems/cmakepackage.rst:
```diff
@@ -0,0 +1,258 @@
+
+   $ cmake .. -G Ninja
+   $ ninja
+   $ ninja install
```
`archer` and `llvm-openmp-ompt` are the only examples currently in Spack. In the case of these packages, it is the developers who recommend using Ninja. From my understanding, Ninja is faster for very large projects? I don't think we need to take a stance on this as a package manager, I'm fine with just using whatever the developers recommend.

citibeth replied:
I would add that info to the docs.

adamjstewart commented on lib/spack/docs/build_systems/cudapackage.rst:
```diff
@@ -0,0 +1,38 @@
+
+Different from other packages, ``CudaPackage`` does not represent a build
+system. Instead its goal is to simplify and unify usage of ``CUDA`` in other
+packages.
```
Yes, this is more of a mixin than a build system. I would like to eventually move this to a separate `mixin` directory outside of `build_systems`.

citibeth replied:
OK please explain that in the docs.

adamjstewart commented on lib/spack/docs/build_systems/custompackage.rst:
```diff
@@ -0,0 +1,202 @@
+.. code-block:: python
+
+   def bootstrap(self, spec, prefix):
+       bootstrap = Executable('./bootstrap')
```
This is the build system used to install `cmake`, not a package that installs _with_ `cmake`.

adamjstewart commented on lib/spack/docs/build_systems/custompackage.rst:
```diff
@@ -0,0 +1,202 @@
+
+.. code-block:: python
+
+   @on_package_attributes(run_tests=True)
```
I believe it's a class attribute that is set when the `--test=root` flag is used.

citibeth commented:
@adamjstewart I don't have time to go over replies individually...  I'd give the big picture as follows:

1. My comments are the result of a random reader reading through the docs once.  If I asked questions about something explained later, that might (or might not) indicate a need to rearrange some of the material.  I did not take the time to un-ask questions that were explained later.

2. Where there are more than one common ways to do things, I would suggest you provide examples showing all of them.  Even if you personally prefer some of the ways over others.  That is how I wrote docs in the past, even documenting features I personally wasn't thrilled with but others liked.  In that vein, I also would suggest that in addition to `spack edit` commands, that the docs tell the reader where the relevant file is.  Some people will want to use `spack edit`, some won't.  (The less frequently you do something, the more it makes sense to NOT have a special command for it).

3. In cases I asked a question and you answered it --- that's material that now needs to be incorporated into the docs.

4. If you used material from other places, that needs to be cited.

5. Wikipedia is not the ultimate authority.  I don't think that their categorization of build systems is right for this purpose.

6. When you mention packages that use a build system, see if you can hyperlink to them in the GitHub repo.

adamjstewart commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,295 @@
+
+.. code-block:: python
+
+   build_directory = 'src'
```
Can you be more specific? There are dozens of packages in Spack that have used this for a while.

citibeth replied:
I thought that `build_directory=src` would change the build directory for me in all my phases.  And yet, as you can see, I still need to change the directory myself.  Either there's a bug, and oversight, or I just don't understand how this is supposed to work:

https://github.com/spack/spack/pull/8166

adamjstewart replied:
It only changes the build directory if you don't override the default functions. See the definition of `lib/spack/spack/build_systems/makefile.py`.

citibeth commented:
> It only changes the build directory if you don't override the default
> functions. See the definition of lib/spack/spack/build_systems/makefile.py
> .
>

That should definitely make it into the docs

citibeth commented:
Look at slide 2 here:

https://www.slideshare.net/DanielPfeifer1/cmake-48475415

CMake is really taking over..

citibeth commented:
See "Don't Abuse Automagic Dependencies" https://www.linux.com/news/best-practices-autotools

And here: "Why Automagic Dependencies are Wrong" https://wiki.gentoo.org/wiki/Project:Quality_Assurance/Automagic_dependencies

adamjstewart commented:
Okay, added another swath of changes. By popular demand, I've added links to the `package.py` for each package mentioned in the build system docs. This links to the package on GitHub, which will also tell them the path to that file if they want to view it locally without the use of `spack edit`. Let me know if I missed any packages. In cases where the same package was referenced multiple times (`perl` and `cmake` in Custom Build Systems), I only added a link to the first occurrence.

tgamblin commented:
@adamjstewart: is this going to be done soon?  I think this stuff is really valuable and maybe we could do it in more than one PR...

adamjstewart commented:
@tgamblin It's done as far as I'm concerned. There can always be additional content added later, but I think it's a solid start. The only thing missing is documentation on `with_or_without` in `AutotoolsPackage`.

adamjstewart commented:
Ping ping

scheibelp commented:
I'll look into reviewing and merging this starting on 6/25. I anticipate it will take a day or two.

adamjstewart commented:
Take your time, I'm happy to tweak any sections, just let me know.

scheibelp commented on lib/spack/docs/build_systems/rpackage.rst:
```diff
@@ -0,0 +1,341 @@
+   > update.packages(ask = FALSE)
+
+
+This works great for users who has internet access, but those on an
```
Typo, should read:

> this works great for users who *have* internet access

scheibelp commented on lib/spack/docs/build_systems/sconspackage.rst:
```diff
@@ -0,0 +1,300 @@
+
+In the case of cantera, using ``env_vars=all`` allows us to use
+Spack's compiler wrappers. If you don't see an option related to
+environment variables, try using Spack's compiler wrappers. The full
```
I think a bit more detail would be useful:

> ... try using Spack's compiler wrappers by passing spack_cc, spack_cxx, and spack_fc via the CC, CXX, and FORTRAN arguments, respectively.

scheibelp commented on lib/spack/docs/build_systems/autotoolspackage.rst:
```diff
@@ -0,0 +1,301 @@
+necessary for installation. If this is the case, your package does not
+require any Autotools dependencies.
+
+However, a basic rule of version control systems is to never commit
```
I think this is a bit more detail than is required for a Spack user. IMO you could refocus the conversation:

> Some packages use higher-level specification files (e.g. ` ``configure.ac`` ` and ` ``Makefile.am`` `) to generate a configure script. In this case, Spack will automatically run `autoreconf` to generate the configure script, but you need to add some dependencies...

That also sidesteps issues related to properly citing the .svg file (although that by itself is not a good reason to omit it).

scheibelp commented:
In https://github.com/spack/spack/pull/5015#discussion_r197976181 I forgot to mention that I created https://github.com/spack/spack/issues/8571 so you could reference that in a TODO if you wish.

adamjstewart commented:
Thanks @scheibelp, I'll try to address these comments by the end of the week.

adamjstewart commented:
> I have build the docs and looked through them and the formatting appears generally good. As a side note, do you know how to view the end product of the documentation (i.e. without building the docs yourself)? If I view the files in the diff it doesn't render everything properly.

I don't build them, I just assume I know what they will look like. But if there's anyone who knows a good way to preview .rst files, it would be @svenevs.

adamjstewart commented on lib/spack/docs/build_systems/cmakepackage.rst:
```diff
@@ -0,0 +1,275 @@
+
+Like Autotools, CMake is a build-script generator. Designed by Kitware,
+CMake is a popular up-and-coming build system. It is
+`becoming the dominant build system <https://trends.google.com/trends/explore?date=all&q=autoconf,automake,cmake,scons>`_
```
I can't remember where I found this (maybe @citibeth pointed me to it), but I found a talk on build systems that used these trends to argue this point. Of course, that doesn't mean it's a valid point.

tgamblin replied:
It's considered the "standard" by the C++ tooling committee, at least, and new IDES like CLion are making the choice to only support CMake.  There are still lots of autotools builds out there but I see them less and less.

"Dominant build system" is probably too general -- there are other very big ecosystems and "build system" is vaguely defined.  I would say it's probably the most popular build system for new C, C++, and Fortran projects.

adamjstewart commented:
Addressed most of @scheibelp's comments. Still thinking about some of the comments on AutotoolsPackage and CMakePackage.

svenevs commented:
> I don't build them, I just assume I know what they will look like

Shame on you :stuck_out_tongue:  After the initial build, edits should be fast if you're only changing a couple files.

> But if there's anyone who knows a good way to preview .rst files, it would be @svenevs.

I defer to the (excellent) advice of @alalazo here: https://github.com/spack/spack/pull/5861#issuecomment-338932100

I think you can also `SPHINXOPTS="" make html` if you don't want to edit the `Makefile` but don't recall, and I'm currently revolting against `python` because it gives me `ModuleNotFoundError: No module named '_struct'`.

**However**, fast options for viewing a reasonable rendering would be to first copy the contents of the entire file you want to check (for GitHub PRs you can click "View File").  Then

1. Create a secret gist on `gist.github.com`, create a file called say `README.rst`.  As long as it has `.rst`, GitHub does a reasonable job rendering reStructuredText.
2. Paste the contents into a file and run `rst2html.py your_file.rst > index.html`.

In both cases, you won't get the full effect (e.g. references to other parts of the docs).  It's best to render it with sphinx.  Additionally, the Sphinx parser and writer is different than both `docutils` (the `rst2html.py` script origin), and GitHub.  So the only way to know for sure is to build them yourself.

Don't forget you can also `make html -j 10` for parallel (but that usually doesn't help much...).  I hope one of those two is marginally useful!

adamjstewart commented:
Just remembered that you can click "View" for a specific file to see the .rst rendered properly. Of course, it won't process TOCs or refs, but it's a good start.

> Don't forget you can also make html -j 10 for parallel 

I'm pretty sure that doesn't help. Once upon a time, the Makefile explicitly contained logic to run Sphinx in parallel, but it secretly wasn't being used because `sphinxcontrib.programoutput` doesn't declare whether or not it is parallel-safe. I tried making it parallel-safe one time, but literally everything broke. I don't think multiple files can cross-reference each other if you want things to build in parallel.

scheibelp commented:
> Still thinking about some of the comments on AutotoolsPackage and CMakePackage.

Is this still true? I wanted to make sure I wasn't letting this slide

adamjstewart commented:
@scheibelp Thanks for checking in!

> Is this still true? I wanted to make sure I wasn't letting this slide

It's not _not_ true. I'm still on the fence on your remaining comments. If anyone else agrees, I can make the change, otherwise I think I'll leave it as is.

tgamblin commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,299 @@
+Make has multiple types of assignment operators. Some Makefiles
+use ``=`` to assign variables. The only way to override these
+variables is to edit the Makefile or override them on the command-line.
+However, Makefiles that use ``?=`` for assignment honor environment
```
Can you add a link to https://www.gnu.org/software/make/manual/make.html#Setting for `?=`?

tgamblin commented on lib/spack/docs/build_systems/makefilepackage.rst:
```diff
@@ -0,0 +1,299 @@
+Variables to watch out for
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The following is a list of common variables to watch out for:
```
Is it worth adding a link to https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html in here somewhere?

citibeth commented:
Thank you Adam

On Tue, Jul 17, 2018 at 8:54 PM Stephen McDowell <notifications@github.com>
wrote:

> This is why I refer to him as SpackMan. He may not be the hero we
> deserve... But definitely the one we need ❤️
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/5015#issuecomment-405774316>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd93OLvgcaJz0ubReehe-6LXPRqmzks5uHodHgaJpZM4OwBuN>
> .
>
